api/eth_types.go

```diff
@@ -0,0 +1,442 @@
+package api
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	mathbig "math/big"
+	"strconv"
+	"strings"
+
+	"github.com/ipfs/go-cid"
+	"github.com/multiformats/go-multihash"
+	"github.com/multiformats/go-varint"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/big"
+	builtintypes "github.com/filecoin-project/go-state-types/builtin"
+	"github.com/filecoin-project/go-state-types/builtin/v10/eam"
+
+	"github.com/filecoin-project/lotus/build"
+)
+
+type EthUint64 uint64
+
+func (e EthUint64) MarshalJSON() ([]byte, error) {
+	if e == 0 {
+		return json.Marshal("0x0")
+	}
+	return json.Marshal(fmt.Sprintf("0x%x", e))
+}
+
+func (e *EthUint64) UnmarshalJSON(b []byte) error {
+	var s string
+	if err := json.Unmarshal(b, &s); err != nil {
+		return err
+	}
+	parsedInt, err := strconv.ParseUint(strings.Replace(s, "0x", "", -1), 16, 64)
+	if err != nil {
+		return err
+	}
+	eint := EthUint64(parsedInt)
+	*e = eint
+	return nil
+}
+
+type EthBigInt big.Int
+
+var (
+	EthBigIntZero = EthBigInt{Int: big.Zero().Int}
+)
+
+func (e EthBigInt) MarshalJSON() ([]byte, error) {
+	if e.Int == nil {
+		return json.Marshal("0x0")
+	}
+	return json.Marshal(fmt.Sprintf("0x%x", e.Int))
+}
+
+func (e *EthBigInt) UnmarshalJSON(b []byte) error {
+	var s string
+	if err := json.Unmarshal(b, &s); err != nil {
+		return err
+	}
+
+	replaced := strings.Replace(s, "0x", "", -1)
+	if len(replaced)%2 == 1 {
+		replaced = "0" + replaced
+	}
+
+	i := new(mathbig.Int)
+	i.SetString(replaced, 16)
+
+	*e = EthBigInt(big.NewFromGo(i))
+	return nil
+}
+
+type EthBytes []byte
+
+func (e EthBytes) MarshalJSON() ([]byte, error) {
+	if len(e) == 0 {
+		return json.Marshal("0x00")
+	}
+	s := hex.EncodeToString(e)
+	if len(s)%2 == 1 {
+		s = "0" + s
+	}
+	return json.Marshal("0x" + s)
+}
+
+func (e *EthBytes) UnmarshalJSON(b []byte) error {
+	var s string
+	if err := json.Unmarshal(b, &s); err != nil {
+		return err
+	}
+
+	s = strings.Replace(s, "0x", "", -1)
+	if len(s)%2 == 1 {
+		s = "0" + s
+	}
+
+	decoded, err := hex.DecodeString(s)
+	if err != nil {
+		return err
+	}
+
+	*e = decoded
+	return nil
+}
+
+type EthBlock struct {
+	ParentHash       EthHash    `json:"parentHash"`
+	Sha3Uncles       EthHash    `json:"sha3Uncles"`
+	Miner            EthAddress `json:"miner"`
+	StateRoot        EthHash    `json:"stateRoot"`
+	TransactionsRoot EthHash    `json:"transactionsRoot"`
+	ReceiptsRoot     EthHash    `json:"receiptsRoot"`
+	// TODO: include LogsBloom
+	Difficulty    EthUint64 `json:"difficulty"`
+	Number        EthUint64 `json:"number"`
+	GasLimit      EthUint64 `json:"gasLimit"`
+	GasUsed       EthUint64 `json:"gasUsed"`
+	Timestamp     EthUint64 `json:"timestamp"`
+	Extradata     []byte    `json:"extraData"`
+	MixHash       EthHash   `json:"mixHash"`
```

Comment:
Shouldn't this just be `hash`?

Reply:
https://openethereum.github.io/JSONRPC-eth-module#eth_getblockbynumber

Reply:
I believe so: https://docs.infura.io/infura/networks/ethereum/json-rpc-methods/eth_getblockbynumber.

Reply:
I believe [this](https://ethereum.github.io/execution-apis/api-documentation/) is the official doc and is more up-to-date. It does have a field called `mixHash`, which is different from `hash` (block hash). It seems that `hash` is missing and I'll add it to the struct. 

Same for the [go-ethereum implementation](https://github.com/ethereum/go-ethereum/blob/c4a662176ec11b9d5718904ccefee753637ab377/core/types/block.go#L84)